File: back-end/src/main/java/es/ull/project/application/usecase/container/UpdateContainerUseCase.java
package es.ull.project.application.usecase.container;

import es.ull.project.domain.entity.Container;
import es.ull.project.domain.enumerate.ServiceZone;
import es.ull.project.domain.enumerate.WasteType;
import es.ull.project.domain.valueobject.capacity.ContainerCapacityLiters;
import es.ull.project.domain.valueobject.demand.DailyWasteDemandLitersPerDay;
import es.ull.project.domain.valueobject.location.Location;
import es.ull.project.domain.valueobject.name.Name;
import java.util.UUID;

/**
 * Use case for updating a container.
 */
public interface UpdateContainerUseCase {
    /**
     * Updates an existing container with the specified attributes.
     *
     * @param id             the unique identifier of the container
     * @param newLocation    the new location
     * @param newWasteType   the new waste type
     * @param newCapacityLiters the new container capacity in liters
     * @param newDailyDemandLitersPerDay the new daily demand in liters per day
     * @param newServiceZone the new service zone
     * @return the updated container
     */
    Container update(UUID id, Name newName, Location newLocation, WasteType newWasteType, ContainerCapacityLiters newCapacityLiters, DailyWasteDemandLitersPerDay newDailyDemandLitersPerDay, ServiceZone newServiceZone);
}

File: front-end/src/application/usecase/ContainerManagement/update-container-use-case.ts
// UpdateContainerUseCase.ts
// Use case contract for updating a container

import type { Either, DataError } from '@ull-tfg/ull-tfg-typescript';
import { Container } from '../../../domain/entity/container';
import { Location } from '../../../domain/valueobject/location/location';
import { WasteType } from '../../../domain/enumerate/waste-type';
import { WasteDemand } from '../../../domain/valueobject/demand/waste-demand';
import { ServiceZone } from '../../../domain/enumerate/service-zone';

/**
 * Use case for updating an existing container in the system.
 * Input: containerId and fields to update
 * Output: Container (entity)
 */

// Command object for input data
export interface UpdateContainerCommand {
    containerId: string;
    updatedFields: Partial<{
        location: Location;
        wasteType: WasteType;
        wasteDemand: WasteDemand;
        serviceZone: ServiceZone | null;
    }>;
}

// Result type for the use case
export type UpdateContainerResult = Container;

// Use case contract
export interface UpdateContainerUseCase {
    /**
     * Handles the update of an existing container
     *
     * @param command Data needed to update the container
     * @returns Either a DataError or the updated Container
     */
    execute(command: UpdateContainerCommand): Promise<Either<DataError, UpdateContainerResult>>;
}
